README.md
# react-native-stylable

  [![Build Status](https://travis-ci.org/vovkasm/react-native-stylable.svg?branch=master)](https://travis-ci.org/vovkasm/react-native-stylable)
  [![Coverage Status](https://coveralls.io/repos/github/vovkasm/react-native-stylable/badge.svg?branch=master)](https://coveralls.io/github/vovkasm/react-native-stylable?branch=master)

Cascading styles for ReactNative.
A components for easy styling you React Native application.

:warning: The package is currently in alpha stage of development. If you find a bug or missing functionality, please feel free to report, but better fix or implement what you want and send a pull request to GitHub repository.

## Installation

```sh
npm install --save react-native-stylable
```

## Usage

```javascript
// app.js
import React from 'react'
import { StyleProvider } from 'react-native-stylable'

import AppView from './AppView'
import styleSheet from './styles'

const App = () => <StyleProvider styleSheet={styleSheet}>
  <AppView/>
</StyleProvider>

export default App
```

```javascript
// styles.js
import { Stylesheet } from 'react-native-stylable'

const s = new Stylesheet()
// Global styles, theme code, etc...
// Will overwrite defaults...
s.addRules({
  'baseText': {
    style: { fontSize: 8 }
  }
})
export default s
```

```javascript
// some component, Item.js for ex
import PropTypes from 'prop-types'
import React from 'react'
import { Image, Text, View } from 'react-native'
import { stylable } from 'react-native-stylable'

import styleSheet from './styles'

// default styles for our component
styleSheet.addDefaultRules({
  Item: {
    style: {
      borderWidth: 1,
      borderColor: 'rgb(240,240,240)',
      margin: 4,
      flex: 1
    }
  },
  'Item Description': {
    style: {
      marginHorizontal: 12
    }
  },
  'Item PriceText': {
    mixins: ['baseText'],
    style: {
      fontWeight: 'bold'
    }
  },
  'Item DescriptionText': {
    mixins: ['baseText'],
    style: {
      marginVertical: 6
    }
  },
  'Item Image': {
    style: {
      left: 0,
      right: 0,
      height: 160
    }
  }
})

const Description = stylable('Description')(View)
const DescriptionText = stylable('DescriptionText')(Text)
const Image = stylable('Image')(Image)
const PriceText = stylable('PriceText')(Text)

class Item extends React.Component {
  static propTypes = {
    modelItem: PropTypes.object.isRequired
  }
  render () {
    const { modelItem, ...other } = this.props
    const price = modelItem.price + modelItem.currency
    const uri = modelItem.imageUri
    return <View {...other}>
      <Image source={{uri}} />
      <Description>
        <DescriptionText>{modelItem.name}</DescriptionText>
        <PriceText>{price}</PriceText>
      </Description>
    </View>
  }
}

export default stylable('Item')(Item)
```

## API

```javascript
import { StyleSheet, StyleProvider, stylable } from 'react-native-stylable'
```

### Stylesheet

Holds style rules.

#### Constructor

Creates new empty style sheet.

```javascript
const styleSheet = new StyleSheet()
```

#### addDefaultRules(rules:Object)

Add default (low priority) rules to style sheet. This method intended to be called from custom components code to
provide default styles.

#### addRules(rules:Object)

Add normal rules. This method intended to be called from application startup to provide customizations.

#### Rules object

Each rule has selector and properties.

Selector is an element name or sequence of names separated by space
(same as descendant selectors in CSS). Also any element name can be suffixed with variant with dot symbol
(ex: `Button.active ButtonTitle`), variants works like classes in CSS.

Properties is an object, that can contains `mixins`, `props` and `style` keys.
 * `mixins` - Array of others element names.
 * `props` - Properties to be added to React component.
 * `style` - Styles to be added to `style` of React component.

### StyleProvider

React component that provides StyleSheet to other stylable components.

`<StyleProvider styleSheet={styleSheet}>{other app components}</StyleSheet>`

### stylable(name:String):Function

Return function (`fn(component:Component):Component`) to produce Higher Order Component with `name` that
pass props and style from StyleSheet to wrapped component. Also generated component can accept one additional
property - `variant` to dynamically change styles.

## Author

[Vladimir Timofeev](https://github.com/vovkasm)

## License

Source code is licensed under the MIT License.


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="0.2.2"></a>
## [0.2.2](https://github.com/vovkasm/react-native-stylable/compare/v0.2.1...v0.2.2) (2017-11-17)


### Bug Fixes

* update precompiled version ([7ca2b49](https://github.com/vovkasm/react-native-stylable/commit/7ca2b49))



<a name="0.2.1"></a>
## [0.2.1](https://github.com/vovkasm/react-native-stylable/compare/v0.2.0...v0.2.1) (2017-11-17)


### Bug Fixes

* Allow to wrap host components in stylable ([f89fc5e](https://github.com/vovkasm/react-native-stylable/commit/f89fc5e))
* Remove hack that no more work with React 16+ ([89528e0](https://github.com/vovkasm/react-native-stylable/commit/89528e0))



<a name="0.2.0"></a>
# [0.2.0](https://github.com/vovkasm/react-native-stylable/compare/v0.1.0...v0.2.0) (2017-08-09)


### Bug Fixes

* fix getting proper name of component ([f037042](https://github.com/vovkasm/react-native-stylable/commit/f037042))


### Features

* package only needed files ([a9e55a1](https://github.com/vovkasm/react-native-stylable/commit/a9e55a1))
* Use better naming for components ([8b063b3](https://github.com/vovkasm/react-native-stylable/commit/8b063b3))



<a name="0.1.0"></a>
# [0.1.0](https://github.com/vovkasm/react-native-stylable/compare/0.0.5...v0.1.0) (2017-06-24)


### Bug Fixes

* make getDisplayName to be almost equal to react getElementName ([6449d80](https://github.com/vovkasm/react-native-stylable/commit/6449d80))
* migrate to prop-types module (React.PropTypes is deprecated) ([1cd5750](https://github.com/vovkasm/react-native-stylable/commit/1cd5750))
* react warn if we try to get ref of pure component ([ffadc1e](https://github.com/vovkasm/react-native-stylable/commit/ffadc1e))


### Features

* **package:** compile & release es5 version ([0e0ba9a](https://github.com/vovkasm/react-native-stylable/commit/0e0ba9a))



# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

# 0.0.5 (2016-12-26)

* Proxy setNativeProps to wrapped components so they can be used as direct Touchable childs. (cb5ba87)
* Add benchmarks for selectors (be0ab83)
* Split bench.js to runner and benchmark function. (e274907)

# 0.0.4 (2016-12-22)

* Use default for React Native .babelrc file. Using babel-plugin-istanbul breaks testing in apps, that use this module. (41ac903)
* Add test for "out-of-tree" components (b07dcee)
* Add additional tests for rules order. (aea0375)

# 0.0.3 (2016-12-21)

The first working version

bench/bench.js

import Benchmark from 'benchmark'

import Stylesheet from '../src/stylesheet'
import Node from '../src/node'

const s = new Stylesheet()
s.addRules({
  stdFont: {style: {fontFamily: 'Helvetica'}},
  pdLeft: {style: {paddingLeft: 2}},
  pdRight: {style: {paddingRight: 2}},
  pdHorizontal: {mixins: ['pdLeft', 'pdRight']},
  baseText: {mixins: ['stdFont', 'pdHorizontal'], style: {fontSize: 12}},
  baseText2: {mixins: ['stdFont'], style: {fontSize: 11}},
  Button: {
    props: {
      color: '#bf7650'
    }
  }
})
s.addDefaultRules({
  'Rule1': {},
  'Rule2': {},
  'Rule3': {},
  'Rule4': {},
  'Rule5': {},
  'Rule6': {},
  'Rule7': {},
  'Rule8': {},
  'Rule9': {},
  'Rule10': {}
})
s.addDefaultRules({
  AppView: {
    style: {
      backgroundColor: '#ffffff',
      justifyContent: 'center',
      alignItems: 'stretch',
      paddingTop: 20,
      flexGrow: 1
    }
  }
})
s.addDefaultRules({
  'AppView Item': {
    style: {
      borderWidth: 1,
      borderColor: 'rgb(240,240,240)',
      margin: 4,
      flex: 1
    }
  },
  'AppView Item Description': {
    style: {
      marginHorizontal: 12
    }
  },
  'AppView Item PriceText': {
    mixins: ['baseText'],
    style: {
      fontWeight: 'bold'
    }
  },
  'AppView Item Text': {
    mixins: ['baseText'],
    style: {
      marginVertical: 6
    }
  },
  'AppView Item Image': {
    style: {
      left: 0,
      right: 0,
      height: 160
    }
  }
})
s.addDefaultRules({
  'Rule1 Text': {style: {flex: 1}},
  'Rule2 Text': {style: {flex: 1}},
  'Rule3 Text': {style: {flex: 1}},
  'Rule4 Text': {style: {flex: 1}},
  'Rule5 Text': {style: {flex: 1}},
  'Rule6 Text': {style: {flex: 1}},
  'Rule7 Text': {style: {flex: 1}}
})

const appNode = new Node('App', {}, undefined, s)
const appViewNode = new Node('AppView', {}, appNode, s)
const elemNode = new Node('Elem', {}, appViewNode, s)
const itemNode = new Node('Item', {}, elemNode, s)
const textNode = new Node('Text', {prop1: 1}, itemNode, s)

const suite = new Benchmark.Suite()
let props
suite.add('new theme & props', function () {
  props = s.getProps(textNode)
  if (props.style.fontSize !== 12) {
    throw new Error('fontSize != 12')
  }
})
suite.on('cycle', function (e) {
  console.log(String(e.target))
  if (e.target.error) {
    console.log('Error: ', e.target.error)
  }
})
suite.on('complete', function () {
  console.log('Done')
})
suite.run()
